pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,852 @@
+			if err != nil {
+				klog.ErrorS(err, "Failed to update PacketCapture status")
+			}
+		}
+	}
+	defer cleanup()
+	pc, err := c.packetCaptureLister.Get(name)
+	if err != nil {
+		return err
+	}
+	srcIP, dstIp, err := c.parseIPs(pc)
+	if err != nil {
+		return err
+	}
+	// checked before, ensured it's not nil here.
+	device := c.getTargetCaptureDevice(pc)
+	klog.V(2).InfoS("Prepare capture on current node", "name", pc.Name, "device", *device)
+	packetCapturer, err := c.newPacketCapturerFn(pc)
+	if err != nil {
+		klog.ErrorS(err, "Failed to create packet source")
+		return err
+	}
+	pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	timeout := defaultTimeoutDuration
+	if pc.Spec.Timeout != nil {
+		timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+	}
+	go func() error {
+		captureErr := c.performCapture(pc, pcState, *device, srcIP, dstIp, timeout, packetCapturer)
+		func() {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			c.captures[name] = packetCapturePhaseCompleted
+			c.numRunningCaptures -= 1
+			c.cond.Signal()
+		}()
+
+		updateErr := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", captureErr)
+		if updateErr != nil {
+			klog.ErrorS(updateErr, "Failed to update PacketCapture status")
+		}
+		return captureErr
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(
+	pc *crdv1alpha1.PacketCapture,
+	captureState *packetCaptureState,
+	device string,
+	srcIP, dstIP net.IP,
+	timeout time.Duration,
+	capturer PacketCapturer) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	packets, err := capturer.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+
+	for {
+		select {
+		case packet := <-packets:
+			if captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum {
+				return nil
+			}
+			captureState.capturedPacketsNum++
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = captureState.pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packets: %w", err)
+			}
+			klog.V(5).InfoS("Capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			reachTarget := captureState.capturedPacketsNum == captureState.targetCapturedPacketsNum
+			// use rate limiter to reduce the times we need to update status.
+			if reachTarget || captureState.updateRateLimiter.Allow() {
+				pc, err := c.packetCaptureLister.Get(captureState.name)
+				if err != nil {
+					return fmt.Errorf("get PacketCapture failed: %w", err)
```

Comment:
This means the PacketCapture is deleted, I think the caller should handle the error differently to skip status update and return nil at the end?

Reply:
in the updateStatus function there is also a exist check in the beginning.